CMakeLists.txt

```diff
@@ -431,8 +431,31 @@ endif()
 
 if(DEFINED BUILD_VERSION)
   set(build_version_argument "-DBUILD_VERSION=${BUILD_VERSION}")
-elseif(EXISTS ${ZEPHYR_BASE}/.git/index)
+elseif(IS_DIRECTORY ${ZEPHYR_BASE}/.git)
   set(git_dependency ${ZEPHYR_BASE}/.git/index)
+elseif(EXISTS ${ZEPHYR_BASE}/.git)
+  # Likely a git-submodule. Let's ask git where the real database is located.
+  find_package(Git QUIET)
+  if(GIT_FOUND)
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} rev-parse --git-dir
```

Comment:
this will cause `git` to be invoked on each CMake re-run, which is unnecessary.

When doing it this way, we should follow the common CMake pattern of creating a `ZEPHYR_GIT_INDEX` cache variable which can be set to `-NOTFOUND` when not found.

This will ensure that `git` is only invoked on first CMake invocation, or on builds where `BUILD_VERSION` is not set and the Zephyr project has not been cloned using `git`.

Something like this.
```
if(DEFINED BUILD_VERSION)
  set(build_version_argument "-DBUILD_VERSION=${BUILD_VERSION}")
elseif(NOT ZEPHYR_GIT_INDEX)
  if(EXISTS ${ZEPHYR_BASE}/.git/index)
    set(ZEPHYR_GIT_INDEX ${ZEPHYR_BASE}/.git/index CACHE PATH "Path to Zephyr git repository index file" FORCE)
  elseif(...)
     ... <git revparse proposal> ...
    set(ZEPHYR_GIT_INDEX ${zephyr_git_dir}/index CACHE PATH "Path to Zephyr git repository index file" FORCE)
  else()
    set(ZEPHYR_GIT_INDEX ZEPHYR_GIT_INDEX-NOTFOUND CACHE PATH "Path to Zephyr git repository index file")
    message(WARNING "ZEPHYR_BASE=${ZEPHYR_BASE} doesn't appear to be a git "
      "repository, please specify '-DBUILD_VERSION=<version>'")
  endif()
endif()

if(ZEPHYR_GIT_INDEX)
  set(git_dependency ${ZEPHYR_GIT_INDEX})
endif()
```

**Note**: A CMake variable having a value ending with the `-NOTFOUND` suffix has a special meaning.

Reply:
This is true. I was debating whether or not the more complex code change is worth it, since the `git` call is probably pretty quick unless your OS's `fork` is terribly slow. But this is obviously fair as well.

So I've done that in the new force pushed version. While reading the code again, I noticed that I had not set `git_dependency` when `stderr` was non-empty. In the new version I **only** discard the `stdout` data from the git call if the program's return code is non-zero (and therefore indicates failure). Now if it returns success and still puts data on `stderr`, you get the warning, but the data on `stdout` is still used.

Reply:
> since the git call is probably pretty quick

For this particular case, I agree there will be little gain.

This is more on a general level and fitting with existing CMake pattern makes things more consistent.